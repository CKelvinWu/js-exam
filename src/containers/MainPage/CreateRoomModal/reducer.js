import { CREATE_ROOM } from './constants';

const initialState = {
  createRoom: false,
  createRoomSuc: false,
  createRoomErr: false,
  createdRoomData: {},
};

function createPageReducer(state = initialState, action) {
  switch (action.type) {
    case CREATE_ROOM.REQUEST:
      return {
        createRoom: true,
        createRoomSuc: false,
        createRoomErr: false,
      };

    case CREATE_ROOM.SUCCESS:
      return {
        createRoom: false,
        createRoomSuc: true,
        createRoomErr: false,
        createdRoomData: {
          ...action.data,
        },
      };

    case CREATE_ROOM.FAILURE:
      return {
        createRoom: false,
        createRoomSuc: false,
        createRoomErr: true,
      };

    case CREATE_ROOM.RESET:
      return {
        createRoom: false,
        createRoomSuc: false,
        createRoomErr: false,
      };

    default:
      return state;
  }
}

export default createPageReducer;
