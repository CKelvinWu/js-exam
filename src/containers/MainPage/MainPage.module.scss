.content {
  display: flex;
  justify-content: center;
  padding: 20px;
}

.column.createRoom {
  height: calc(100vh - 48px);
  width: 50%;
  text-align: center;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 50px;
}
