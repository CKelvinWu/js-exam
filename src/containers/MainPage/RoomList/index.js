import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { List, Avatar } from 'antd';
import style from './RoomList.module.scss';

const RoomList = ({ rooms, isLoading, signedOn, hostings }) => (
  <List
    itemLayout="horizontal"
    dataSource={rooms}
    loading={isLoading}
    renderItem={room => (
      <Link
        to={{
          pathname: `/admin/dispatch/${room.id}`,
        }}
      >
        <List.Item
          style={{ borderBottom: '1px solid #ddd' }}
          className={style.listItem}
        >
          <List.Item.Meta
            className={style.listMeta}
            avatar={<Avatar icon="home" className={style.avatar} />}
            title={
              <div>
                <div className={style.roomNameHeader}>Room </div>
                <div className={style.roomName}>{room.description}</div>
              </div>
            }
            description={
              <div>
                <div className={style.subjectIdHeader}>Candidate </div>
                <div className={style.subjectId}>{room.subjectId}</div>
              </div>
            }
          />
          {signedOn && hostings.includes(room.id) && (
            <div className={style.listContent}>Host</div>
          )}
        </List.Item>
      </Link>
    )}
  />
);

RoomList.propTypes = {
  rooms: PropTypes.array,
  isLoading: PropTypes.bool,
  signedOn: PropTypes.bool,
  hostings: PropTypes.array,
};

export default RoomList;
