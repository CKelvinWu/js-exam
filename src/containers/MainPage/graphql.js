export const listRooms = `query ListRooms(
    $filter: ModelRoomFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listRooms(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        subjectId
        description
        createTime
      }
      nextToken
    }
  }`;

export const onCreateRoom = `subscription OnCreateRoom {
    onCreateRoom {
        id
        subjectId
        description
        createTime
    }
  }`;

export const onDeleteRoom = `subscription OnDeleteRoom {
    onDeleteRoom {
      id
    }
  }`;
