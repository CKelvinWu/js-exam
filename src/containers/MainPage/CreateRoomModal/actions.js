/*
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */
import { API, graphqlOperation } from 'aws-amplify';
import * as mutations from 'graphql/mutations';
import { action } from 'utils/actionsHelper';
import { message } from 'antd';
import { setHostings } from 'redux/login/actions';
import { CREATE_ROOM } from './constants';

export const createRoomActions = {
  reset: () => action(CREATE_ROOM.RESET),
  request: data => action(CREATE_ROOM.REQUEST, { data }),
  success: data => action(CREATE_ROOM.SUCCESS, { data }),
  failure: error => action(CREATE_ROOM.FAILURE, { error }),
};

const createRoomQl = `mutation CreateRoom($input: CreateRoomInput!) {
  createRoom(input: $input) {
    id
    test {
      id
      subjectId
      description
      timeBegin
      timeEnd
      status
      tags
      host{
        id
        name
      }
    }
    subjectId
    description
    host {
      id
      name
    }
    createTime
    password
    users {
      items {
        id
        name
      }
      nextToken
    }
    currentRecord {
      id
      subjectId
      syncCode
      timeBegin
      timeEnd
      status
    }
  }
}
`;

export function createRoom(data) {
  return async dispatch => {
    dispatch(createRoomActions.request(data));
    try {
      const roomNum = Math.floor(Math.random() * 98) + 1;
      const roomChar = String.fromCharCode(Math.floor(Math.random() * 26) + 65);
      const createTime = new Date();
      // TODO: change graphql schema such that a jeUser can host mutiple room and test
      const { data: jeUserData } = await API.graphql(
        graphqlOperation(mutations.createJeUser, {
          input: {
            name: localStorage.username,
          },
        }),
      );
      const { data: testData } = await API.graphql(
        graphqlOperation(mutations.createTest, {
          input: {
            timeBegin: createTime,
            status: 'open',
            testHostId: jeUserData.createJEUser.id,
            // temporarily store host name in tags, since a jeUser cannot host mutiple test in current graphql schema
            tags: localStorage.username,
            ...data,
          },
        }),
      );
      const { data: roomData } = await API.graphql(
        graphqlOperation(createRoomQl, {
          input: {
            roomTestId: testData.createTest.id,
            description: roomChar + roomNum,
            createTime,
            ...data,
          },
        }),
      );
      // TODO: change graphql schema to connect jeUser and test and room. So that don't have to update manually.
      const { data: jeUserUpdate } = await API.graphql(
        graphqlOperation(mutations.updateJeUser, {
          input: {
            id: jeUserData.createJEUser.id,
            jEUserRoomId: roomData.createRoom.id,
            jEUserTestId: testData.createTest.id,
            jEUserHostTestId: testData.createTest.id,
          },
        }),
      );
      const { data: testUpdate } = await API.graphql(
        graphqlOperation(mutations.updateTest, {
          input: {
            id: testData.createTest.id,
            testRoomId: roomData.createRoom.id,
          },
        }),
      );

      dispatch(createRoomActions.success(roomData.createRoom));

      /* add a hosting room into 'hostings' array in local storage */
      dispatch(setHostings(roomData.createRoom.id));
    } catch (error) {
      dispatch(createRoomActions.failure(error));
      message.error('Create room failed');
      console.log(error);
    }
  };
}
