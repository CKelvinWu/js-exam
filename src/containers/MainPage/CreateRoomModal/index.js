import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';

import injectReducer from 'utils/injectReducer';

import { Modal, Button, Input, message, Icon } from 'antd';

import { setRoomHost } from 'redux/room/actions';
import { REDUCER_KEY } from 'redux/createRoomModal/constants';
import { createRoomActions, createRoom } from 'redux/createRoomModal/actions';
import reducer from 'redux/createRoomModal/reducer';
import style from './CreateRoomModal.module.scss';

class CreateRoomModal extends React.PureComponent {
  state = {
    name: '',
    disabled: true,
  };

  constructor(props) {
    super(props);
    this.examLinkInput = React.createRef();
  }

  handleInputOnChange = e => {
    this.setState({
      name: e.target.value,
      disabled: e.target.value === '',
    });
  };

  handleOnOK = () => {
    const { name } = this.state;

    this.props.onCreateRoom({
      subjectId: name,
    });
  };

  handleOnAfterClose = () => {
    this.setState({
      name: '',
      disabled: true,
    });
    this.props.onReset();
  };

  copyLink = () => {
    this.examLinkInput.current.select();
    document.execCommand('copy');
    message.success('Link has been copied.');
  };

  toRoom = () => {
    this.handleOnAfterClose();
    const { history, createdRoomData, onSetRoomHost } = this.props;
    onSetRoomHost(true);
    history.push(`/admin/dispatch/${createdRoomData.id}`);
  };

  render() {
    const {
      visible,
      onClose,
      isCreatingRoom,
      createRoomSuc,
      createdRoomData,
    } = this.props;

    const { name, disabled } = this.state;

    let title = 'Create a room';
    let okButtonLabel = isCreatingRoom ? 'Creating...' : 'Create';
    let onOK = this.handleOnOK;

    if (createRoomSuc) {
      title = `Welcome to Room - ${createdRoomData.description}`;
      okButtonLabel = 'Enter Room';
      onOK = this.toRoom;
    }

    return (
      <Modal
        visible={visible}
        maskClosable={false}
        title={title}
        closable={!isCreatingRoom}
        okText={okButtonLabel}
        cancelButtonProps={{
          disabled: isCreatingRoom,
        }}
        okButtonProps={{
          disabled: disabled || isCreatingRoom,
        }}
        onOk={onOK}
        onCancel={onClose}
        afterClose={this.handleOnAfterClose}
      >
        {!createRoomSuc ? (
          <Input
            placeholder="Please enter Candidate name"
            prefix={<Icon type="user" />}
            value={name}
            onChange={this.handleInputOnChange}
          />
        ) : (
          <React.Fragment>
            <p>Candidate: {createdRoomData.subjectId}</p>
            <div className={style.content}>
              <Input
                size="large"
                readOnly
                value={`${window.location.origin}${
                  window.location.pathname
                }${window.location.hash.replace('admin', 'exam')}/${
                  createdRoomData.id
                }`}
                ref={this.examLinkInput}
              />
              <Button className={style.copyButton} onClick={this.copyLink}>
                <Icon type="copy" />
              </Button>
            </div>
          </React.Fragment>
        )}
      </Modal>
    );
  }
}

CreateRoomModal.propTypes = {
  visible: PropTypes.bool,
  history: PropTypes.object.isRequired,
  onClose: PropTypes.func.isRequired,
  onReset: PropTypes.func.isRequired,
  onCreateRoom: PropTypes.func,
  onSetRoomHost: PropTypes.func,
  isCreatingRoom: PropTypes.bool,
  createRoomSuc: PropTypes.bool,
  createdRoomData: PropTypes.object,
};

const mapDispatchToProps = dispatch => ({
  onReset: () => dispatch(createRoomActions.reset()),
  onCreateRoom: data => dispatch(createRoom(data)),
  onSetRoomHost: data => dispatch(setRoomHost(data)),
});

const mapStateToProps = state => ({
  isCreatingRoom: state[REDUCER_KEY].createRoom,
  createRoomSuc: state[REDUCER_KEY].createRoomSuc,
  createdRoomData: state[REDUCER_KEY].createdRoomData,
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: REDUCER_KEY, reducer });

export default compose(withReducer, withConnect)(CreateRoomModal);
