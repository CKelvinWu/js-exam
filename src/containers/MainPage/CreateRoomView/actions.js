/*
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */
import { API, graphqlOperation } from 'aws-amplify';
import * as mutations from 'graphql/mutations';
import { action } from 'utils/actionsHelper';
import { message } from 'antd';
import { setHostings } from 'redux/login/actions';
import { CREATE_ROOM } from './constants';

export const createRoomActions = {
  request: data => action(CREATE_ROOM.REQUEST, { data }),
  success: data => action(CREATE_ROOM.SUCCESS, { data }),
  failure: error => action(CREATE_ROOM.FAILURE, { error }),
};

export function createRoom(data) {
  return async dispatch => {
    dispatch(createRoomActions.request(data));
    try {
      const roomNum = Math.floor(Math.random() * 98) + 1;
      const roomChar = String.fromCharCode(Math.floor(Math.random() * 26) + 65);
      const createTime = new Date();
      const { data: testData } = await API.graphql(
        graphqlOperation(mutations.createTest, {
          input: {
            timeBegin: createTime,
            status: 'open',
            ...data,
          },
        }),
      );
      const { data: roomData } = await API.graphql(
        graphqlOperation(mutations.createRoom, {
          input: {
            roomTestId: testData.createTest.id,
            description: roomChar + roomNum,
            createTime,
            ...data,
          },
        }),
      );
      dispatch(createRoomActions.success(roomData.createRoom));

      /* add a hosting room into 'hostings' array in local storage */
      dispatch(setHostings(roomData.createRoom.id));
    } catch (error) {
      dispatch(createRoomActions.failure(error));
      message.error('Create room failed');
      console.log(error);
    }
  };
}
