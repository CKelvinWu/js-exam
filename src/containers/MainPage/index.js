import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Connect } from 'aws-amplify-react';
import { connect } from 'react-redux';
import { graphqlOperation } from 'aws-amplify';
import { Button, Input, Modal } from 'antd';
import debounce from 'lodash/debounce';

import { deleteRoomAction } from 'redux/room/actions';
import { listRooms } from 'graphql/queries';
import { onCreateRoom } from 'graphql/subscriptions';

import PageEmpty from 'components/PageEmpty';
import PageSpin from 'components/PageSpin';
import PageControlBar from 'components/PageControlBar';
import RoomList from './RoomList';
import CreateRoomModal from './CreateRoomModal';

import style from './MainPage.module.scss';

const Search = Input.Search;

class MainPage extends Component {
  state = {
    searchKeyword: '',
    isModalVisible: false,
    delConfirmModalVisible: false,
    delRoom: null,
    delRoomList: [],
  };

  handleOnSearch = debounce(value => {
    this.setState({
      searchKeyword: value.toLowerCase(),
    });
  }, 300);

  handleNewModalButtonOnClick = () => {
    this.setState({
      isModalVisible: true,
    });
  };

  handleModalOnClose = () => {
    this.setState({
      isModalVisible: false,
    });
  };

  hideDelConfirmModal = () => {
    this.setState({
      delConfirmModalVisible: false,
    });
  };

  handleOnOkDelConfirmModal = async () => {
    const { delRoom, delRoomList } = this.state;
    const { deleteRoomAction } = this.props;
    await deleteRoomAction(delRoom);
    this.setState({
      delRoomList: [...delRoomList, delRoom.id],
    });

    this.hideDelConfirmModal();
  };

  render() {
    const { history, signedOn, hostings } = this.props;
    const {
      isModalVisible,
      searchKeyword,
      delConfirmModalVisible,
      delRoom,
      delRoomList,
    } = this.state;

    return (
      <div className={style.Mainpage}>
        <PageControlBar>
          <div>
            <Button
              type="primary"
              icon="plus"
              onClick={this.handleNewModalButtonOnClick}
            >
              Create a room
            </Button>
          </div>
          <div>
            <Search
              style={{ width: 400 }}
              placeholder="input search text"
              onChange={e => this.handleOnSearch(e.target.value)}
            />
          </div>
        </PageControlBar>
        <div className={`${style.content}`}>
          {/* TODO: Room list with with lazy-loading next dataset. Here we load 1000 rooms instead. */}
          <Connect
            query={graphqlOperation(listRooms, { limit: 1000 })}
            subscription={graphqlOperation(onCreateRoom)}
            onSubscriptionMsg={(prev, { onCreateRoom: createdRoom }) => {
              prev.listRooms.items.unshift(createdRoom);
              return prev;
            }}
          >
            {({ data: { listRooms: rooms }, loading, error }) => {
              const outputRooms =
                rooms &&
                rooms.items
                  .map(room => {
                    room.createTimeByDate = new Date(room.createTime);
                    return room;
                  })
                  .sort((a, b) => {
                    return b.createTimeByDate - a.createTimeByDate;
                  })
                  .filter(room => {
                    return (
                      (room.subjectId.toLowerCase().includes(searchKeyword) ||
                        room.description
                          .toLowerCase()
                          .includes(searchKeyword)) &&
                      !delRoomList.includes(room.id)
                    );
                  });

              return (
                <PageSpin spinning={loading}>
                  {!loading && error && (
                    <PageEmpty description={<span>Error Occuring</span>} />
                  )}

                  {!loading && !outputRooms.length && (
                    <PageEmpty
                      description={<span>Room Not Found</span>}
                      image="default"
                    />
                  )}

                  {!loading && outputRooms.length && (
                    <RoomList
                      rooms={outputRooms}
                      signedOn={signedOn}
                      hostings={hostings}
                      triggerDelRoom={room => {
                        this.setState({
                          delConfirmModalVisible: true,
                          delRoom: room,
                        });
                      }}
                    />
                  )}
                </PageSpin>
              );
            }}
          </Connect>
        </div>

        <CreateRoomModal
          visible={isModalVisible}
          history={history}
          onClose={this.handleModalOnClose}
        />
        <Modal
          title=""
          visible={delConfirmModalVisible}
          okType="danger"
          okText="Delete"
          onOk={this.handleOnOkDelConfirmModal}
          onCancel={this.hideDelConfirmModal}
        >
          Are you sure you want to delete room{' '}
          <b>{delRoom ? delRoom.description : ''}</b> ?
        </Modal>
      </div>
    );
  }
}

MainPage.propTypes = {
  history: PropTypes.object.isRequired,
  signedOn: PropTypes.bool,
  hostings: PropTypes.array,
};

const mapStateToProps = state => ({
  signedOn: !!state.login.username,
  hostings: state.login.hostings,
});

const mapDispatchToProps = dispatch => ({
  deleteRoomAction: delRoom => dispatch(deleteRoomAction(delRoom)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MainPage);
