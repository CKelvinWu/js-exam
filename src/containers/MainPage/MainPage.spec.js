import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux/configureStore';
import { deleteRoomAction } from 'redux/room/actions';

import MainPage from './index';
import RoomList from './RoomList';
import CreateRoomModal from './CreateRoomModal';

jest.mock('./CreateRoomModal', () => jest.fn(() => <div>CreateRoomModal</div>));

jest.mock('./RoomList', () => jest.fn(() => <div>RoomList</div>));

jest.mock('aws-amplify-react', () => {
  const rooms = {
    items: [
      {
        id: 'f509ee49-c4d6-4d3c-b357-2d18454b01f2',
        subjectId: 'test room 1',
        description: 'Z77',
        createTime: '2020-01-31T07:07:03.295Z',
      },
      {
        id: 'f509ee49-c4d6-4d3c-b357-2d18454b01f2',
        subjectId: 'test room 1',
        description: 'Z77',
        createTime: '2020-01-31T07:07:03.295Z',
      },
      {
        id: 'c6f86538-3661-4cd3-a8e2-09578958bd39',
        subjectId: 'test room 2',
        description: 'V72',
        createTime: '2020-01-31T07:06:41.936Z',
      },
      {
        id: 'cfb2f8c4-8a2d-442d-a836-467be88a85f5',
        subjectId: 'test room 3',
        description: 'I87',
        createTime: '2020-02-13T06:24:08.206Z',
      },
    ],
  };
  const data = {
    listRooms: rooms,
    id: 'c6f86538-3661-4cd3-a8e2-09578958bd39',
  };
  const arg = { data, loading: false, mutatuion: jest.fn() };
  return {
    Connect: props => props.children(arg),
  };
});

jest.mock('redux/room/actions', () => ({
  deleteRoomAction: jest.fn(data => data),
}));

const rooms = [
  {
    id: 'cfb2f8c4-8a2d-442d-a836-467be88a85f5',
    subjectId: 'test room 3',
    description: 'I87',
    createTime: '2020-02-13T06:24:08.206Z',
    createTimeByDate: new Date('2020-02-13T06:24:08.206Z'),
  },
  {
    id: 'f509ee49-c4d6-4d3c-b357-2d18454b01f2',
    subjectId: 'test room 1',
    description: 'Z77',
    createTime: '2020-01-31T07:07:03.295Z',
    createTimeByDate: new Date('2020-01-31T07:07:03.295Z'),
  },
];

describe('<MainPage />', () => {
  let wrapper;
  let subject;
  let instance;
  let dispatch;
  const initialState = {
    searchKeyword: '',
    isModalVisible: false,
    delConfirmModalVisible: false,
    delRoom: null,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    const store = configureStore({
      login: {
        username: 'user name',
        hostings: [],
      },
    });
    store.dispatch = jest.fn(action => action);
    const props = {
      history: {},
    };
    wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <MainPage {...props} />
        </BrowserRouter>
      </Provider>,
    );
    subject = wrapper.find('MainPage');
    instance = subject.instance();
    dispatch = wrapper.prop('store').dispatch;
  });

  it('should render properly', () => {
    expect(wrapper.length).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
    expect(subject.length).toBeTruthy();
    expect(subject.find('PageControlBar').length).toBeTruthy();
    expect(subject.find('Connect').length).toBe(2);
    expect(RoomList).toBeCalledWith(
      {
        hostings: [],
        rooms,
        signedOn: true,
        triggerDelRoom: expect.any(Function),
      },
      {},
    );
    expect(CreateRoomModal).toBeCalledWith(
      {
        visible: false,
        history: subject.prop('history'),
        onClose: instance.handleModalOnClose,
      },
      {},
    );
  });

  it('should handle on search', done => {
    jest.useRealTimers();
    jest.spyOn(instance, 'handleOnSearch');
    const searchInput = subject.find('input');
    searchInput.simulate('change', { target: { value: 'Test Room 1' } });
    expect(instance.handleOnSearch).toBeCalled();
    expect(subject.state()).toEqual(initialState);
    setTimeout(() => {
      expect(subject.state()).toEqual({
        ...initialState,
        searchKeyword: 'test room 1',
      });
      expect(RoomList).lastCalledWith(
        {
          hostings: [],
          rooms: [rooms[1]],
          signedOn: true,
          triggerDelRoom: expect.any(Function),
        },
        {},
      );
      done();
    }, 325);
  });

  it('should handle new modal button on click', () => {
    const spy = jest.spyOn(instance, 'handleNewModalButtonOnClick');
    instance.forceUpdate();
    const createRoomButton = subject.find('button');
    createRoomButton.simulate('click');
    expect(instance.handleNewModalButtonOnClick).toBeCalled();
    expect(subject.state()).toEqual({ ...initialState, isModalVisible: true });
    spy.mockRestore();
  });

  it('should handle modal on close', () => {
    subject.setState({ isModalVisible: true });
    instance.handleModalOnClose();
    expect(subject.state()).toEqual({ ...initialState, isModalVisible: false });
  });

  it('should handle hide delete confirm modal', () => {
    subject.setState({ delConfirmModalVisible: true });
    instance.hideDelConfirmModal();
    expect(subject.state()).toEqual({
      ...initialState,
      delConfirmModalVisible: false,
    });
  });

  it('should handle on ok del confirm modal', async () => {
    jest.spyOn(instance, 'hideDelConfirmModal');
    subject.setState({ delRoom: rooms[0] });
    await instance.handleOnOkDelConfirmModal();
    expect(deleteRoomAction).toBeCalledWith(rooms[0]);
    expect(dispatch).toBeCalledWith(rooms[0]);
    expect(instance.hideDelConfirmModal).toBeCalled();
  });

  it('should handle trigger del room', () => {
    const triggerDelRoom = RoomList.mock.calls[0][0].triggerDelRoom;
    triggerDelRoom(rooms[0]);
    expect(subject.state()).toEqual({
      ...initialState,
      delConfirmModalVisible: true,
      delRoom: rooms[0],
    });
  });

  it('should handle graphql subscription', () => {
    // onCreateRoomMsg might change the input array
    const testRooms = [...rooms];
    const onCreateRoomMsg = subject
      .find('Connect')
      .first()
      .prop('onSubscriptionMsg');
    let result = onCreateRoomMsg(
      { listRooms: { items: testRooms } },
      { onCreateRoom: { id: 'cfb2f8c4-8a2d-442d-a836-467be88a85f5' } },
    );
    expect(result.listRooms.items).toEqual(rooms);
    result = onCreateRoomMsg(
      { listRooms: { items: testRooms } },
      { onCreateRoom: { id: 'newId' } },
    );
    expect(result.listRooms.items).toEqual([{ id: 'newId' }, ...rooms]);

    const onDeleteRoomMsg = subject
      .find('Connect')
      .last()
      .prop('onSubscriptionMsg');
    result = onDeleteRoomMsg(rooms, { onDeleteRoom: { id: 'delId' } });
    expect(result).toEqual({ id: 'delId' });
  });

  it('should render properly when no data or error', () => {
    const aws = require('aws-amplify-react');
    aws.Connect = jest.fn(props =>
      props.children({
        data: { listRooms: { items: [] } },
        loading: false,
        mutatuion: jest.fn(),
      }),
    );
    subject.setState({});
    wrapper.update();
    let PageEmpty = wrapper.find('PageEmpty');
    expect(PageEmpty.length).toBeTruthy();
    expect(PageEmpty.text()).toBe('Room Not Found');

    aws.Connect = jest.fn(props =>
      props.children({
        data: { listRooms: { items: rooms } },
        error: 'error',
        loading: false,
        mutatuion: jest.fn(),
      }),
    );
    subject.setState({});
    wrapper.update();
    PageEmpty = wrapper.find('PageEmpty');
    expect(PageEmpty.length).toBeTruthy();
    expect(PageEmpty.text()).toBe('Error Occuring');
  });
});
