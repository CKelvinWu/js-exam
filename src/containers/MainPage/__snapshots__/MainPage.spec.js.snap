// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> should render properly 1`] = `"<div class=\\"Mainpage\\"><div class=\\"control\\"><div><button type=\\"button\\" class=\\"ant-btn ant-btn-primary\\"><i aria-label=\\"icon: plus\\" class=\\"anticon anticon-plus\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"plus\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z\\"></path><path d=\\"M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z\\"></path></svg></i><span>Create a room</span></button></div><div><span class=\\"ant-input-search ant-input-affix-wrapper\\" style=\\"width: 400px;\\"><input placeholder=\\"input search text\\" class=\\"ant-input\\" type=\\"text\\" value=\\"\\"><span class=\\"ant-input-suffix\\"><i aria-label=\\"icon: search\\" tabindex=\\"-1\\" class=\\"anticon anticon-search ant-input-search-icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"search\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z\\"></path></svg></i></span></span></div></div><div class=\\"content\\"><div class=\\"ant-spin-nested-loading\\"><div class=\\"ant-spin-container\\"><div>RoomList</div></div></div></div><div>CreateRoomModal</div></div>"`;
